src/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/FileBasedPrograms/VirtualProjectXmlProvider.cs

```diff
@@ -0,0 +1,92 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Immutable;
+using System.Composition;
+using System.Runtime.InteropServices;
+using System.Security;
+using System.Security.Cryptography;
+using System.Text;
+using System.Text.Json;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.Extensions.Logging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.LanguageServer.FileBasedPrograms;
+
+[Export(typeof(VirtualProjectXmlProvider)), Shared]
+[method: ImportingConstructor]
+[method: Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+internal class VirtualProjectXmlProvider(DotnetCliHelper dotnetCliHelper, ILoggerFactory loggerFactory)
+{
+    private readonly ILogger<VirtualProjectXmlProvider> _logger = loggerFactory.CreateLogger<VirtualProjectXmlProvider>();
+
+    internal async Task<(string VirtualProjectXml, ImmutableArray<SimpleDiagnostic> Diagnostics)?> GetVirtualProjectContentAsync(string documentFilePath, CancellationToken cancellationToken)
+    {
+        var workingDirectory = Path.GetDirectoryName(documentFilePath);
+        var process = dotnetCliHelper.Run(["run-api"], workingDirectory, shouldLocalizeOutput: true, redirectStandardInput: true);
+
+        cancellationToken.Register(() =>
+        {
+            process?.Kill();
+        });
+
+        var input = new RunApiInput.GetProject() { EntryPointFileFullPath = documentFilePath };
+        var inputJson = JsonSerializer.Serialize(input, RunFileApiJsonSerializerContext.Default.RunApiInput);
+        await process.StandardInput.WriteAsync(inputJson);
+        process.StandardInput.Close();
+
+        process.ErrorDataReceived += (sender, args) => _logger.LogDebug($"dotnet run-api: {args.Data}");
+        process.BeginErrorReadLine();
+
+        var responseJson = await process.StandardOutput.ReadLineAsync(cancellationToken);
+        await process.WaitForExitAsync(cancellationToken);
+
+        if (string.IsNullOrWhiteSpace(responseJson))
+        {
+            _logger.LogError($"dotnet run-api returned exit code '{process.ExitCode}' but did not return a response.");
+            return null;
+        }
+
+        var response = JsonSerializer.Deserialize(responseJson, RunFileApiJsonSerializerContext.Default.RunApiOutput);
```

Comment:
do we want to catch errors here?  Looks like it is currently throwing in the tests because it has some other error output.  

Perhaps if the process returns a non-zero exit code we shouldn't even try to deserialize?